src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java

```diff
@@ -183,13 +189,141 @@ private void commitEntriesToStorage(Map<WALEntry, WALEntry> entries) {
     }
   }
 
+  /**
+   * Extract the latest offset and file path from the WAL file.
+   * Attempt with the most recent WAL file, and fall back to the old file if it's not applicable.
+   *
+   * <p> The old WAL is used when the most recent WAL file has already been truncated
+   * and is not existent, which may happen when the connector has flushed all records,
+   * and is restarted. </p>
+   *
+   * @return the latest offset and filepath from the WAL file or null
+   */
+  public Pair<Long, String> extractLatestOffsetFromWAL() {
+    String oldWALFile = logFile + OLD_LOG_EXTENSION;
+    try {
+      Pair<Long, String> latestOffset = null;
+      if (storage.exists(logFile)) {
+        log.trace("Restoring offset from WAL file: {}", logFile);
+        if (reader == null) {
+          reader = newWalFileReader();
+        } else {
+          // reset read position after apply()
+          reader.seekToFirstRecord();
+        }
+        List<String> committedFileBatch = getLastFilledBlockFromWAL(reader);
+        // At this point the committedFilenames list will contain the
+        // filenames in the last BEGIN-END block of the file. Find the latest offsets among these.
+        latestOffset = getLatestOffsetFromList(committedFileBatch);
+      }
+
+      // attempt to use old log file if recent WAL is empty or non-existent
+      if (latestOffset == null && storage.exists(oldWALFile)) {
```

Comment:
Does it make sense to attempt this part if the above part throws IOException?

Reply:
It wouldn't for the following reason: the most recent WAL will have the most recent offsets, and falling back to the old WAL would cause to use an older offset (restore to an older point than the files which are in HDFS). If the recent `WAL` exists and throws an IOException the fallback should be the recursive file search, not the old `WAL`. Let me know if that makes sense.